src/transforms/mod.rs

```diff
@@ -0,0 +1,396 @@
+    let mut iter = vector_pair_values.into_iter();
+    let v1 = iter.next().unwrap().into_values();
+    let v2 = iter.next().unwrap().into_values();
+
+    Ok((v1, v2))
+}
+
+fn value_and_gradient(
+    value_and_grad: mlx_closure_value_and_grad,
+    arrays: impl Iterator<Item = impl AsRef<Array>>,
+) -> Result<(Vec<Array>, Vec<Array>), Exception> {
+    let input_vector = VectorArray::from_iter(arrays);
+
+    let vector_pair = unsafe {
+        let c_vector_pair = try_catch_c_ptr_expr! {
+            mlx_closure_value_and_grad_apply(value_and_grad, input_vector.as_ptr())
+        };
+        VectorVectorArray::from_ptr(c_vector_pair)
+    };
+
+    let vector_pair_values: SmallVec<[VectorArray; 2]> = vector_pair.into_values();
+    let mut iter = vector_pair_values.into_iter();
+    let values_vec = iter.next().unwrap().into_values();
+    let gradients_vec = iter.next().unwrap().into_values();
+
+    Ok((values_vec, gradients_vec))
+}
+
+fn build_gradient<'a, F>(
+    f: F,
+    argument_numbers: &'a [i32],
+) -> impl FnMut(&[Array]) -> Result<Vec<Array>, Exception> + 'a
+where
+    F: FnMut(&[Array]) -> Vec<Array> + 'a,
+{
+    let closure = Closure::new(f);
+    move |arrays: &[Array]| -> Result<Vec<Array>, Exception> {
+        unsafe {
+            let c_value_and_grad = try_catch_c_ptr_expr! {
+                mlx_sys::mlx_value_and_grad(
+                    closure.as_ptr(),
+                    argument_numbers.as_ptr(),
+                    argument_numbers.len(),
+                )
+            };
+
+            let result = value_and_gradient(c_value_and_grad, arrays.iter())?;
+            Ok(result.1)
+        }
+    }
+}
+
+fn build_value_and_gradient<'a, F>(
+    f: F,
+    argument_numbers: &'a [i32],
+) -> impl FnMut(&[Array]) -> Result<(Vec<Array>, Vec<Array>), Exception> + 'a
+where
+    F: FnMut(&[Array]) -> Vec<Array> + 'a,
+{
+    let closure = Closure::new(f);
+    move |arrays: &[Array]| {
+        let c_value_and_grad = unsafe {
+            try_catch_c_ptr_expr! {
+                mlx_sys::mlx_value_and_grad(
+                    closure.as_ptr(),
+                    argument_numbers.as_ptr(),
+                    argument_numbers.len(),
+                )
+            }
+        };
+
+        value_and_gradient(c_value_and_grad, arrays.iter())
+    }
+}
+
+fn clone_by_increment_ref_count(src: &Array) -> Array {
+    unsafe {
+        mlx_retain(src.as_ptr() as *mut _);
+        Array::from_ptr(src.as_ptr())
+    }
+}
+
+/// Returns a function which computes the value and gradient of `f`.
+pub fn value_and_grad<'a, F>(
+    f: F,
+    argument_numbers: &'a [i32],
+) -> impl FnMut(&[Array]) -> Result<(Vec<Array>, Vec<Array>), Exception> + 'a
+where
+    F: FnMut(&[Array]) -> Vec<Array> + 'a,
+{
+    build_value_and_gradient(f, argument_numbers)
+}
+
+/// Similar to `value_and_grad`, but unwraps the result.
+pub fn value_and_grad_unwrapped<'a, F>(
```

Comment:
we haven't really done `unwrapped` so far, is this specifically necessary for something? Otherwise I'd wait just a bit before adding them just here. Same for the other instances of `unwrapped` in this file.

Reply:
It was originally added so that one could nest `grad()` to compute 2nd order gradient, but there's still lifetime issue. So I guess we could remove them.

Reply:
Resolved in [1083a60](https://github.com/oxideai/mlx-rs/pull/26/commits/1083a60eb69b8c2dc37ae125bd90b0d65627d6c8)